tensorflow-framework/tensorflow-data.md

```diff
@@ -0,0 +1,209 @@
+
+Note: An `Output` is a symbolic handle to a tensor produced by a TensorFlow op. In graph
+mode, `Output` objects will not have a concrete `Tensor` value unless all dependent operations
+are run in a `Session` (this is done "in-real-time" in eager mode).
+
+### Construction
+
+Datasets can be constructed either directly from a data source (e.g. a list of tensors representing the components of the dataset), or as a transformation on an existing dataset.
+
+#### From Data Source
+
+To construct a dataset from a list of tensor components, use 
+`Dataset.fromTensorSlices( ... )`. For example, say we are working
+with a standard feature/label dataset which has 4 elements.
+
+```java
+FloatNdArray features = StdArrays.ndCopyOf(
+        new float[][] {
+        {1, 2, 3},
+        {4, 5, 6},
+        {7, 8, 9},
+        {10, 11, 12}
+});
+
+FloatNdArray labels = NdArrays.vectorOf(0, 1, 1, 0);
+```
+
+A dataset can be constructed from a list of the constant `Operand`s generated
+from this dataset, and a list of `DataType` objects corresponding
+to the type of each component:
+
+Note: Each of the input components must share the same first "batch" dimension.
+
+```java
+Ops tf = // ... TensorFlow Ops accessor (either graph or eager)
+Dataset dataset = Dataset.fromTensorSlices(
+    Arrays.asList(tf.constant(features), tf.constant(labels)),
+    Arrays.asList(TInt32.DTYPE, TInt32.DTYPE)
+);
+```
+
+
+Other data sources are also possible, using `tf.data` ops; these include TFRecord files, CSV files, and more.
+
+#### Transformations on Existing Datasets
+
+Once a dataset has been created from a data source it can be transformed by calling
+methods on the `Dataset` object. For example, to group elements in the above dataset into batches of size `2`, use `Dataset.batch(int batchSize)`:
+
+```java
+dataset = dataset.batch(2)
+```
+
+Each dataset transformation alters both the values and shapes of the original
+ elements, and returns a *new* `Dataset` object.
+
+In this case, the original dataset had 4 elements of shape `[features: (3,) labels: (1,)]`.
+Once the `.batch` transformation is applied, the new dataset has 2 elements (batches) of shape `[features: (2, 3), labels: (2, 1)]`.
+
+Similar transformations include `.skip`, `.take`, `.map`, `.filter`, etc.
+
+
+### Iterating over Dataset Elements
+
+The primary use of a dataset is for iteration over its elements.
+Each row (or batch) element is represented as a list of tensor components, with
+type `List<Output<?>>`. The tensor components of this element can be accessed using `List.get(int index)`.
+
+It is recommended to use `Tensor.expect(DataType<?> dtype)` to restore types
+to the retrieved tensors.
+
+#### Using DatastetIterator
+The `DatasetIterator` class provides abstractions for creating and using
+iterators in graph and eager mode. These will be explained here; however
+end-users should only interact with `Iterator` objects through the methods
```

Comment:
I think many reference to `Iterator` should be replaced by `DatasetIterator` here.

Reply:
Thanks, changed these!

Reply:
Once we get a stable docs location, we should probably go through this file and make the class names link to the javadoc for that class. Nothing to worry about in this PR but something to keep in mind for later.

Reply:
Yes, definitely!

Reply:
Yup, definitely!